jujugui/static/gui/src/app/components/user-profile/model-list/model-list.js

```diff
@@ -108,6 +110,134 @@ YUI.add('user-profile-model-list', function() {
     },
 
     /**
+      Display the confirmation for destroying a model.
+
+      @method _displayConfirmation
+      @param {String} model the model to destroy.
+    */
+    _displayConfirmation: function(model) {
+      this.setState({destroyModel: model});
+    },
+
+    /**
+      Generate the confirmation for destroying a model.
+
+      @method _displayConfirmation
+      @return {Object} the confirmation component.
+    */
+    _generateConfirmation: function() {
+      const model = this.state.destroyModel;
+      const addNotification = this.props.addNotification;
+      if (!model) {
+        return;
+      }
+      if (model.name === 'controller') {
+        addNotification({
+          title: 'Cannot destroy model',
+          message: 'The controller model cannot be destroyed. ',
+          level: 'error'
+        });
+        return;
+      }
+      const buttons = [{
+        title: 'Cancel',
+        action: this._displayConfirmation.bind(this, null),
+        type: 'base'
+      }, {
+        title: 'Destroy',
+        action: this._destroyModel,
+        type: 'destructive'
+      }];
+      const name = model.name.split('/')[1];
+      const message = `Are you sure you want to destroy ${name}?`
+        + ' All the applications and units included in the model will be'
+        + ' destroyed. This action cannot be undone.';
+      return (
+        <juju.components.ConfirmationPopup
+          buttons={buttons}
+          message={message}
+          title="Destroy model" />);
+    },
+
+    /**
+      Makes a request of JIMM or JES to delete a selected model.
+
+      @method _destroyModel
+    */
+    _destroyModel: function() {
+      const model = this.state.destroyModel;
+      const uuid = model.uuid;
+      // Hide the confirmation popup.
+      this._displayConfirmation(null);
+      const xhr = this.props.destroyModels([uuid], this._destroyModelCallback);
+      this.xhrs.push(xhr);
+    },
+
+    /**
+      Callback for the JIMM and JES delete model call.
+
+      @method _destroyModelCallback
+      @param {String} err The error from the request, or null.
+      @param {Array} modelList The list of models.
+    */
+    _destroyModelCallback: function(err, results) {
+      const addNotification = this.props.addNotification;
+      // Handle global errors.
+      if (err) {
+        addNotification({
+          title: 'Model destruction failed',
+          message: 'The model failed to be destroyed: ' + err,
+          level: 'error'
+        });
+        return;
+      }
+      // Check for bad data.
+      let uuids;
+      try {
```

Comment:
I agree that it's good to be defensive when receiving data from the network, but I think here we can safely assume that results is an object, and that it includes a field with the model uuid. If controllerAPI.destroyModels does not feel safe enough, then please let's add the defensive code there rather than here. The API callback returns an error, and once the error is handled by API callers, they can also assume everything else is ok.